README.md
# zreion

[![build](https://github.com/plaplant/zreion/workflows/Run%20Tests/badge.svg?branch=main)](https://github.com/plaplant/zreion/actions)
[![coverage](https://codecov.io/gh/plaplant/zreion/badge.svg?branch=main)](https://codecov.io/gh/plaplant/zreion)
[![license](https://img.shields.io/github/license/plaplant/zreion)](https://opensource.org/licenses/MIT)

`zreion` is a way of quickly computing a "redshift of reionization" field using
a semi-numeric method developed in [Battaglia et
al. (2013)](https://ui.adsabs.harvard.edu/abs/2013ApJ...776...81B/abstract). The
method assumes the redshift of reionization for a particular point in a
cosmological volume is a biased tracer of the matter field, and can be written
as a parameterized bias function. For a full derivation and comparison with
simulations, see the linked paper.

## Installation

Installing the package can be performed by checking the repo out, and then running:

```
pip install .
```

Dependencies should be handled automatically by `pip` if not already installed.

## Dependencies

* setuptools >= 38.3
* cython
* numpy
* pyfftw

## Tests

The repo includes a test suite, which can be invoked by running `pytest` in the
top-level of the repo. In addition to the package dependencies, it requires
`packaging`, `pytest-cov` and `pytest-cases`. These can be installed using:

```
pip install .[test]
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.2.2] - 2022-07-29

### Fixed
- Fixed PyPI upload yet again.

## [0.2.1] - 2022-07-29

### Fixed
- Fixed GitHub action for PyPI upload.

## [0.2.0] - 2022-07-29

### Added
- Allow users to specify `b0` parameter in scale-dependent bias term.

## [0.1.0] - 2020-01-25

Initial release


tests/__init__.py
# -*- coding: utf-8 -*-
# Copyright (c) 2021 Paul La Plante
# Licensed under the MIT License

"""Initialize test suite."""


tests/conftest.py
# -*- coding: utf-8 -*-
"""
conftest.py for zreion.

This space intentionally left blank.
"""


tests/test_zreion.py
# -*- coding: utf-8 -*-
# Copyright (c) 2021 Paul La Plante
# Licensed under the MIT License
"""Tests for zreion.zreion module."""

import pytest
import pytest_cases
import numpy as np
from packaging.version import parse

import zreion.zreion as zreion


@pytest.fixture
def fake_data_deterministic():
    """Define a fixture for fake data."""
    is_random = False
    data_shape = (8, 8, 8)
    data = np.empty(data_shape, dtype=np.float32)
    for i in range(data_shape[0]):
        for j in range(data_shape[1]):
            for k in range(data_shape[2]):
                data[i, j, k] = 100 * i + 10 * j + k

    return data, is_random


@pytest.fixture
def fake_data_random():
    """Define a fixture for fake data."""
    is_random = True
    data_shape = (8, 8, 8)
    np.random.seed(8675309)
    data = np.asarray(np.random.normal(size=data_shape), dtype=np.float32)

    return data, is_random


@pytest.fixture
def fake_data_deterministic_anisotropic():
    """Define a fixture for fake data."""
    is_random = False
    data_shape = (13, 11, 9)
    data = np.empty(data_shape, dtype=np.float32)
    for i in range(data_shape[0]):
        for j in range(data_shape[1]):
            for k in range(data_shape[2]):
                data[i, j, k] = i + 1.5 * j - k

    return data, is_random


@pytest.fixture
def fake_data_random_anisotropic():
    """Define a fixture for fake data."""
    is_random = True
    data_shape = (13, 11, 9)
    np.random.seed(8675309)
    data = np.asarray(np.random.normal(size=data_shape), dtype=np.float32)

    return data, is_random


# define cases decorator
fake_data_cases = pytest_cases.parametrize(
    "fake_data",
    [
        pytest_cases.fixture_ref(fake_data_deterministic),
        pytest_cases.fixture_ref(fake_data_random),
        pytest_cases.fixture_ref(fake_data_deterministic_anisotropic),
        pytest_cases.fixture_ref(fake_data_random_anisotropic),
    ],
)


def test_tophat():
    """Test tophat window function calculation."""
    invals = np.linspace(1, 10, num=50)
    refvals = 3 * (np.sin(invals) - invals * np.cos(invals)) / invals**3
    outvals = zreion.tophat(invals)
    assert np.allclose(refvals, outvals)

    # test negative values
    # function is even, so our same refvals work
    invals = -invals
    outvals = zreion.tophat(invals)
    assert np.allclose(refvals, outvals)

    # also test small values
    invals = np.asarray([-1e-7, -1e-8, -1e-9, 0, 1e-9, 1e-8, 1e-7])
    refvals = 1 - invals**2 / 10.0
    # we get a numpy RuntimeWarning with small/invalid values
    with pytest.warns(RuntimeWarning) as record:
        outvals = zreion.tophat(invals)
    assert len(record) == 1
    # error message changes based on numpy version
    np_ver = parse(np.__version__)
    if np_ver.major == 1 and np_ver.minor < 23:
        assert record[0].message.args[0] == "invalid value encountered in true_divide"
    else:
        assert record[0].message.args[0] == "invalid value encountered in divide"

    # make sure values agree
    assert np.allclose(refvals, outvals)

    return


def test_sinc():
    """Test sinc function calculation."""
    invals = np.linspace(1, 10, num=50)
    refvals = np.sin(invals) / invals
    outvals = zreion.sinc(invals)
    assert np.allclose(refvals, outvals)

    # test negative values
    # function is even, so our same refvals work
    invals = -invals
    outvals = zreion.sinc(invals)
    assert np.allclose(refvals, outvals)

    # also test small values
    invals = np.asarray([-1e-7, -1e-8, -1e-9, 0, 1e-9, 1e-8, 1e-7])
    refvals = 1 - invals**2 / 6.0
    # we get a numpy RuntimeWarning with small/invalid values
    with pytest.warns(RuntimeWarning) as record:
        outvals = zreion.sinc(invals)
    assert len(record) == 1
    # error message changes based on numpy version
    np_ver = parse(np.__version__)
    if np_ver.major == 1 and np_ver.minor < 23:
        assert record[0].message.args[0] == "invalid value encountered in true_divide"
    else:
        assert record[0].message.args[0] == "invalid value encountered in divide"

    # make sure values agree
    assert np.allclose(refvals, outvals)

    return


@fake_data_cases
def test_fft3d_forward(fake_data):
    """Test _fft3d function."""
    data, is_random = fake_data

    # use numpy as reference FFT
    numpy_fft = np.fft.rfftn(data)

    # run zreion fft
    input_shape = data.shape
    zreion_fft = zreion._fft3d(data, input_shape, direction="f")
    # need relatively high tolerance because numpy fft is internally done at
    # double precision, seemingly without a way to change it
    assert np.allclose(numpy_fft, zreion_fft, atol=1e-4)
    assert zreion_fft.dtype.type is np.complex64

    return


@fake_data_cases
def test_fft3d_forward_double_precision(fake_data):
    """Test _fft3d function with double-precision."""
    data, is_random = fake_data
    data = np.asarray(data, dtype=np.float64)

    # use numpy as reference FFT
    numpy_fft = np.fft.rfftn(data)

    # run zreion fft
    input_shape = data.shape
    zreion_fft = zreion._fft3d(data, input_shape, direction="f")
    assert np.allclose(numpy_fft, zreion_fft)
    assert zreion_fft.dtype.type is np.complex128

    return


@fake_data_cases
def test_fft3d_backward(fake_data):
    """Test backward _fft3d function."""
    data, is_random = fake_data
    # use numpy as reference FFT
    fake_data_ft = np.asarray(np.fft.rfftn(data), dtype=np.complex64)

    # run zreion fft
    input_shape = data.shape
    zreion_fft = zreion._fft3d(fake_data_ft, input_shape, direction="b")
    assert np.allclose(zreion_fft, data, atol=1e-6)
    assert zreion_fft.dtype.type is np.float32

    return


@fake_data_cases
def test_fft3d_backward_double(fake_data):
    """Test backward _fft3d function with double-precision."""
    data, is_random = fake_data
    data = np.asarray(data, dtype=np.float64)
    fake_data_ft = np.asarray(np.fft.rfftn(data), dtype=np.complex128)

    # run zreion fft
    input_shape = data.shape
    zreion_fft = zreion._fft3d(fake_data_ft, input_shape, direction="b")
    assert np.allclose(zreion_fft, data)
    assert zreion_fft.dtype.type is np.float64

    return


def test_fft3d_direction_error(fake_data_deterministic):
    """Test _fft3d error in direction kwarg."""
    data, is_random = fake_data_deterministic

    # test giving a bogus direction
    with pytest.raises(ValueError) as cm:
        zreion._fft3d(data, data.shape, direction="blah")
    assert str(cm.value).startswith('"direction" must be "f" or "b", got "blah"')

    return


def test_fft3d_datatype_errors(fake_data_deterministic):
    """Test _fft3d errors with bad datatypes."""
    data, is_random = fake_data_deterministic

    # test using incorrect direction for real data
    with pytest.raises(ValueError) as cm:
        zreion._fft3d(data, data.shape, direction="b")
    assert str(cm.value).startswith("a backward transform requires")

    # test using incorrect direction for complex data
    complex_data = np.asarray(data, dtype=np.complex64)
    with pytest.raises(ValueError) as cm:
        zreion._fft3d(complex_data, data.shape, direction="f")
    assert str(cm.value).startswith("a forward transform requires")

    # test using integer datatype
    int_data = np.asarray(data, dtype=np.int32)
    with pytest.raises(ValueError) as cm:
        zreion._fft3d(int_data, data.shape, direction="f")
    assert str(cm.value).startswith("a forward transform requires")

    return


def test_fft3d_rank_errors(fake_data_deterministic):
    """Test _fft3d errors with data rank."""
    data, is_random = fake_data_deterministic
    with pytest.raises(ValueError) as cm:
        zreion._fft3d(data, (0,), direction="f")
    assert str(cm.value).startswith("data_shape must have 3 dimensions")

    data_1d = data[0, 0, :]
    with pytest.raises(ValueError) as cm:
        zreion._fft3d(data_1d, data.shape, direction="f")
    assert str(cm.value).startswith("input array must be a 3-dimensional array")

    return


@fake_data_cases
def test_apply_zreion(fake_data):
    """Test apply_zreion."""
    data, is_random = fake_data
    # clean up input data
    if not is_random:
        data_mean = np.mean(data)
        data = (data - data_mean) / data_mean

    # define model parameters
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = True
    zre = zreion.apply_zreion(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)

    # make sure results make sense
    if is_random:
        # agreement is not great for random values because of small array size
        atol = 1e-1
    else:
        atol = 1e-6
    assert np.isclose(np.mean(zre), zmean, atol=atol)

    # also test passing in boxsize as a tuple--should give same answer
    boxsize = (boxsize, boxsize, boxsize)
    zre2 = zreion.apply_zreion(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)
    assert np.allclose(zre, zre2)

    return


@fake_data_cases
def test_apply_zreion_no_deconvolution(fake_data):
    """Test apply_zreion with no deconvolution."""
    data, is_random = fake_data
    # clean up input data
    if not is_random:
        data_mean = np.mean(data)
        data = (data - data_mean) / data_mean

    # test with no deconvolution
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = False
    zre = zreion.apply_zreion(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)

    # make sure results make sense
    if is_random:
        # agreement is not great for random values because of small array size
        atol = 1e-1
    else:
        atol = 1e-6
    assert np.isclose(np.mean(zre), zmean, atol=atol)

    return


def test_apply_zreion_errors(fake_data_deterministic):
    """Test errors in the apply_zreion function."""
    data, is_random = fake_data_deterministic

    # define model parameters
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = True

    # test using only a 1d density array
    data_1d = data[0, 0, :]
    with pytest.raises(ValueError) as cm:
        zreion.apply_zreion(data_1d, zmean, alpha, k0, boxsize, rsmooth, deconvolve)
    assert str(cm.value).startswith("density must be a 3d array")

    # test using a bad boxsize
    boxsize = (10.0, 10.0)
    with pytest.raises(ValueError) as cm:
        zreion.apply_zreion(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)
    assert str(cm.value).startswith(
        "boxsize must be either a single number or an array of length 3"
    )

    return


@fake_data_cases
def test_apply_zreion_fast(fake_data):
    """Test apply_zreion_fast."""
    data, is_random = fake_data
    # clean up input data
    if not is_random:
        data_mean = np.mean(data)
        data = (data - data_mean) / data_mean

    # define model parameters
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = True
    zre = zreion.apply_zreion_fast(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)

    # make sure results make sense
    if is_random:
        # agreement is not great for random values because of small array size
        atol = 1e-1
    else:
        atol = 1e-6
    assert np.isclose(np.mean(zre), zmean, atol=atol)

    # also test passing in boxsize as a tuple--should give same answer
    boxsize = (boxsize, boxsize, boxsize)
    zre2 = zreion.apply_zreion_fast(
        data, zmean, alpha, k0, boxsize, rsmooth, deconvolve
    )
    assert np.allclose(zre, zre2)

    return


@fake_data_cases
def test_apply_zreion_fast_no_deconvolution(fake_data):
    """Test apply_zreion_fast with no deconvolution."""
    data, is_random = fake_data
    # clean up input data
    if not is_random:
        data_mean = np.mean(data)
        data = (data - data_mean) / data_mean

    # test with no deconvolution
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = False
    zre = zreion.apply_zreion_fast(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)

    # make sure results make sense
    if is_random:
        # agreement is not great for random values because of small array size
        atol = 1e-1
    else:
        atol = 1e-6
    assert np.isclose(np.mean(zre), zmean, atol=atol)

    return


def test_apply_zreion_fast_errors(fake_data_deterministic):
    """Test errors in the apply_zreion_fast function."""
    data, is_random = fake_data_deterministic

    # define model parameters
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = True

    # test using only a 1d density array
    data_1d = data[0, 0, :]
    with pytest.raises(ValueError) as cm:
        zreion.apply_zreion_fast(
            data_1d, zmean, alpha, k0, boxsize, rsmooth, deconvolve
        )
    assert str(cm.value).startswith("density must be a 3d array")

    # test using a bad boxsize
    boxsize = (10.0, 10.0)
    with pytest.raises(ValueError) as cm:
        zreion.apply_zreion_fast(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)
    assert str(cm.value).startswith(
        "boxsize must be either a single number or an array of length 3"
    )

    return


@fake_data_cases
def test_compare_methods(fake_data):
    """Test that fast and "slow" methods yield same answer."""
    data, is_random = fake_data

    # define model parameters
    zmean = 8.0
    alpha = 0.9
    k0 = 1.0
    boxsize = 10.0
    rsmooth = 1.0
    deconvolve = True

    # use slow function
    zre1 = zreion.apply_zreion(data, zmean, alpha, k0, boxsize, rsmooth, deconvolve)

    # use fast function
    zre2 = zreion.apply_zreion_fast(
        data, zmean, alpha, k0, boxsize, rsmooth, deconvolve
    )

    assert np.allclose(zre1, zre2)

    return


src/zreion/zreion.py
# -*- coding: utf-8 -*-
# Copyright (c) 2020 Paul La Plante
# Licensed under the MIT License

"""Main module for applying zreion function."""

import warnings
import numpy as np
import pyfftw

from . import _zreion


def tophat(x):
    """
    Compute spherical tophat Fourier window function.

    Compute the window function of a Fourier tophat window function W_R(k),
    where R is the characteristic radius of the tophat in real space and k is
    the amplitude of the Fourier vector. As can be shown, the tophat function
    has the form:

        W_R(k) = 3 * (sin(R * k) - (R * k) * cos(R * k)) / (R * k)**3.

    For small values of R*k, we use the Taylor series approximation, which is:

        W_R(k) ~ 1 - (R * k)**2 / 10.

    Note that this function itself does not handle a potential RuntimeWarning
    arising from arguments that cause division by zero. However, np.where
    statement handles this cases appropriately, so it can be generally be
    ignored by the user for typical arguments.

    Parameters
    ----------
    x : ndarray
        An array containing the arguments R*k.

    Returns
    -------
    ndarray
        An array of the same size containing the tophat applied to each element.
    """
    return np.where(
        np.abs(x) > 1e-6, 3 * (np.sin(x) - x * np.cos(x)) / x**3, 1 - x**2 / 10.0
    )


def sinc(x):
    """
    Compute sinc(x) = sin(x) / x function.

    Compute the sinc function sin(x) / x, which is necessary for deconvolving
    the cloud-in-cell (CIC) window function typically used to deposit particles
    onto a grid. For small values of x, we use the Taylor series appriximation,
    which is:

        sinc(x) ~ 1 - x**2 / 6.

    Note that this function itself does not handle a potential RuntimeWarning
    arising from arguments that cause division by zero. However, np.where
    statement handles this cases appropriately, so it can be generally be
    ignored by the user for typical arguments.

    Parameters
    ----------
    x : ndarray
        An array containing the arguments x.

    Returns
    -------
    ndarray
        An array of the same size containing sinc applied to each element.
    """
    return np.where(np.abs(x) > 1e-6, np.sin(x) / x, 1 - x**2 / 6.0)


def _fft3d(array, data_shape, direction="f"):
    """
    Apply an FFT using pyFFTW.

    Parameters
    ----------
    array : ndarray
        The array to apply the transform to.
    data_shape : 3-ple of int
        The shape of the input data array. Used for determining the size of the
        output array for a forward transform and the target shape for a backward
        one.
    direction : str
        The direction of the transform. Must be either "f" (for forward) or "b"
        (for backward).

    Returns
    -------
    ndarray
        An ndarray of the resulting transform.
    """
    if direction.lower() not in ["f", "b"]:
        raise ValueError(f'"direction" must be "f" or "b", got "{direction}"')
    dtype = array.dtype.type
    if direction.lower() == "f":
        if dtype is np.float32:
            precision = "single"
        elif dtype is np.float64:
            precision = "double"
        else:
            raise ValueError(
                "a forward transform requires input have np.float32 or np.float64 "
                "datatype"
            )
    else:  # "b"
        if dtype is np.complex64:
            precision = "single"
        elif dtype is np.complex128:
            precision = "double"
        else:
            raise ValueError(
                "a backward transform requires input to have np.complex64 or "
                "np.complex128 datatype"
            )

    if precision == "single":
        input_dtype = "float32"
        output_dtype = "complex64"
    else:  # "double"
        input_dtype = "float64"
        output_dtype = "complex128"

    # define size of FFTW arrays
    if len(array.shape) != 3:
        raise ValueError("input array must be a 3-dimensional array")
    if len(data_shape) != 3:
        raise ValueError("data_shape must have 3 dimensions")
    padded_shape = (data_shape[0], data_shape[1], 2 * (data_shape[2] // 2 + 1))
    full_array = pyfftw.empty_aligned(
        padded_shape, input_dtype, n=pyfftw.simd_alignment
    )

    # make input and output arrays
    real_input = full_array[:, :, : data_shape[2]]
    complex_output = full_array.view(output_dtype)

    if direction == "f":
        fftw_obj = pyfftw.FFTW(real_input, complex_output, axes=(0, 1, 2))
        real_input[:] = array
    else:
        fftw_obj = pyfftw.FFTW(
            complex_output, real_input, axes=(0, 1, 2), direction="FFTW_BACKWARD"
        )
        complex_output[:] = array

    # perform computation and return
    return fftw_obj()


def apply_zreion(
    density, zmean, alpha, k0, boxsize, rsmooth=1.0, deconvolve=True, b0=1.0 / 1.686
):
    r"""
    Apply zreion ionization history to density field.

    Parameters
    ----------
    density : numpy array
        A numpy array of rank 3 holding the density field. Note this should be
        the overdensity field delta = (rho - <rho>)/<rho>, where <rho> is the
        average density value. This quantity has a mean of 0 and a minimum value
        of -1.
    zmean : float
        The mean redshift of reionization.
    alpha : float
        The alpha value of the bias parameter.
    k0 : float
        The k0 value of the bais parameter, in units of h/Mpc.
    boxsize : float or array of floats
        The physical extent of the box along each dimension. If a single value,
        this is assumed to be the same for each axis.
    rsmooth: float, optional
        The smoothing length of the reionziation field, in Mpc/h. Defaults to 1
        Mpc/h.
    deconvolve : bool, optional
        Whether to deconvolve the CIC particle deposition window. If density
        grid is derived in Eulerian space directly, this step is not needed.
    b0 : float, optional
        The b_0 value of the bias parameter. Defaults to 1 / \delta_c.

    Returns
    -------
    zreion : numpy array
        A numpy array of rank 3 containing the redshift corresponding to when
        that portion of the volume was reionized.
    """
    # check that parameters make sense
    if len(density.shape) != 3:
        raise ValueError("density must be a 3d array")
    if isinstance(boxsize, (int, float, np.float_)):
        boxsize = np.asarray([np.float64(boxsize)])
    else:
        # assume it's an array of length 1 or 3
        if len(boxsize) not in (1, 3):
            raise ValueError(
                "boxsize must be either a single number or an array of length 3"
            )

    # compute FFT of density field
    density_fft = np.fft.rfftn(density)

    # compute spherical k-indices for cells
    nx, ny, nz = density.shape
    if len(boxsize) == 1:
        lx = ly = lz = boxsize[0]
    else:
        lx, ly, lz = boxsize
    dx = lx / (2 * np.pi * nx)
    dy = ly / (2 * np.pi * ny)
    dz = lz / (2 * np.pi * nz)
    kx = np.fft.fftfreq(nx, d=dx)
    ky = np.fft.fftfreq(ny, d=dy)
    kz = np.fft.rfftfreq(nz, d=dz)  # note we're using rfftfreq!

    # compute bias factor
    kkx, kky, kkz = np.meshgrid(kx, ky, kz, indexing="ij")
    spherical_k = np.sqrt(kkx**2 + kky**2 + kkz**2)
    bias_val = b0 / (1 + spherical_k / k0) ** alpha

    # compute smoothing factor
    # turn off numpy errors that come from near-zero values
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        smoothing_window = tophat(spherical_k * rsmooth)

    if deconvolve:
        # compute deconvolution window in grid units
        kkx *= lx / nx
        kky *= ly / ny
        kkz *= lz / nz
        # turn off numpy errors that come from near-zero values
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            deconv_window = (sinc(kkx / 2) * sinc(kky / 2) * sinc(kkz / 2)) ** 2
    else:
        deconv_window = np.ones_like(density_fft, dtype=np.float64)

    assert (
        bias_val.shape == density_fft.shape
    ), "Bias and density grids are not compatible"
    assert (
        smoothing_window.shape == density_fft.shape
    ), "Smoothing window and density grids are not compatible"
    assert (
        deconv_window.shape == density_fft.shape
    ), "Deconvolution window and density grids are not compatible"

    # apply transformations to Fourier field
    density_fft *= bias_val
    density_fft *= smoothing_window
    density_fft /= deconv_window

    # inverse FFT
    zreion = np.fft.irfftn(density_fft, density.shape)

    # finish computing zreion field
    zreion *= 1 + zmean
    zreion += zmean

    return zreion


def apply_zreion_fast(
    density, zmean, alpha, k0, boxsize, rsmooth=1.0, deconvolve=True, b0=1.0 / 1.686
):
    r"""
    Use as a fast, drop-in replacement for apply_zreion.

    The speedups are accomplished by using pyfftw for FFT computation, and
    parallelized cython code for most of the rest of the calculation. As an
    added benefit, this version requires significantly less memory.

    Parameters
    ----------
    density : numpy array
        A numpy array of rank 3 holding the density field. Note this should be
        the overdensity field delta = (rho - <rho>)/<rho>, where <rho> is the
        average density value. This quantity has a mean of 0 and a minimum value
        of -1.
    zmean : float
        The mean redshift of reionization.
    alpha : float
        The alpha value of the bias parameter.
    k0 : float
        The k0 value of the bais parameter, in units of h/Mpc.
    boxsize : float or array of floats
        The physical extent of the box along each dimension. If a single value,
        this is assumed to be the same for each axis.
    rsmooth: float, optional
        The smoothing length of the reionziation field, in Mpc/h. Defaults to 1
        Mpc/h.
    deconvolve : bool, optional
        Whether to deconvolve the CIC particle deposition window. If density
        grid is derived in Eulerian space directly, this step is not needed.
    b0 : float, optional
        The b_0 value of the bias parameter. Defaults to 1 / \delta_c.

    Returns
    -------
    zreion : numpy array
        A numpy array of rank 3 containing the redshift corresponding to when
        that portion of the volume was reionized.
    """
    # check that parameters make sense
    if len(density.shape) != 3:
        raise ValueError("density must be a 3d array")
    if isinstance(boxsize, (int, float, np.float_)):
        boxsize = np.asarray([np.float64(boxsize)])
    else:
        # assume it's an array of length 1 or 3
        if len(boxsize) not in (1, 3):
            raise ValueError(
                "boxsize must be either a single number or an array of length 3"
            )

    # save input type
    input_type = density.dtype

    # unpack boxsize argument
    if len(boxsize) == 1:
        lx = ly = lz = boxsize[0]
    else:
        lx, ly, lz = boxsize

    # save input array shape
    array_shape = density.shape

    # perform fft
    density_fft = _fft3d(density, array_shape, direction="f")

    # cast to required type
    density_fft = density_fft.astype(np.complex64)

    # call cython funtion for applying bias relation
    # need to tell function if last dimension has an odd number of elements
    odd_nz = density.shape[2] % 2 == 1
    density_fft = np.asarray(
        _zreion._apply_zreion(
            density_fft, alpha, k0, lx, ly, lz, rsmooth, deconvolve, odd_nz
        )
    )

    # perform inverse fft
    density = _fft3d(density_fft, array_shape, direction="b")

    # finish computing zreion field
    density *= 1 + zmean
    density += zmean

    return density.astype(input_type)
